pkg/k8shandler/visualization.go

```diff
@@ -821,3 +824,143 @@ func createSharedConfig(namespace, kibanaAppURL, kibanaInfraURL string) *v1.Conf
 		},
 	)
 }
+
+func newKibanaCustomResource(cluster *logging.ClusterLogging, kibanaName string) *es.Kibana {
+	cr := &es.Kibana{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Kibana",
+			APIVersion: es.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kibanaName,
+			Namespace: cluster.Namespace,
+		},
+		Spec: es.KibanaSpec{
+			Image: utils.GetComponentImage("kibana"),
+			Resources: &v1.ResourceRequirements{
+				Limits: v1.ResourceList{
+					v1.ResourceMemory: defaultKibanaMemory,
+				},
+				Requests: v1.ResourceList{
+					v1.ResourceMemory: defaultKibanaMemory,
+					v1.ResourceCPU:    defaultKibanaCpuRequest,
+				},
+			},
+		},
+	}
+
+	utils.AddOwnerRefToObject(cr, utils.AsOwner(cluster))
+	return cr
+}
+
+func (clusterRequest *ClusterLoggingRequest) getKibanaCR() (*es.Kibana, error) {
+	var kb *es.Kibana
+	err := clusterRequest.client.Get(context.TODO(),
+		client.ObjectKey{
+			Namespace: clusterRequest.cluster.Namespace,
+			Name:      "instance",
+		}, kb)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to get kibana")
+	}
+	return kb, nil
+}
+
+func (clusterRequest *ClusterLoggingRequest) createOrUpdateKibanaCR() error {
+	cr := newKibanaCustomResource(clusterRequest.cluster, constants.SingletonName)
+
+	err := clusterRequest.Create(cr)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("unable to create kibana cr. E: %s", err.Error())
+	}
+
+	if clusterRequest.isManaged() {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			return clusterRequest.updateKibanaCRDIfRequired(cr)
+		})
+		if retryErr != nil {
+			return retryErr
+		}
+	}
+
+	return nil
+}
+
+func (clusterRequest *ClusterLoggingRequest) updateKibanaCRDIfRequired(cr *es.Kibana) error {
+	current := &es.Kibana{}
+
+	if err := clusterRequest.Get(cr.Name, current); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get kibana CR: %s", err.Error())
+	}
+
+	if isKibanaCRDDifferent(current, cr) {
+		if err := clusterRequest.Update(current); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Check if CRD is not equal to desired. If it's not, returns true and
+// updates the CURRENT crd to be equal to DESIRED
+func isKibanaCRDDifferent(current *es.Kibana, desired *es.Kibana) bool {
```

Comment:
i'm assuming since its passing the pointer to this when we update `current` in this function the changes are available outside the function?

Reply:
Yes, that's correct. Since we passing a pointer, there is no need to return it again from the function.